### PR TITLE
Updated collision to detect ladder and doors

### PR DESCRIPTION
**Pull request by beserkzombie**

### CODE REVIEW

File: Overworld.gd [hunk 1]
```diff
@@ -28,6 +28,11 @@ func place_player():
 func _process(delta):
 	counter += delta
 	$CanvasLayer/Counter.text = "Counter: " + str(counter)
+	var cpos = $TileMap.world_to_map($Player.position)
```
Comment (demwilson) [inline Overworld.gd]: What does `world_to_map` do?

File: Player.gd [hunk 2]
```diff
@@ -21,6 +21,8 @@ onready var animation_player = $AnimationPlayer
 onready var animation_tree = $AnimationTree
 onready var animation_state = animation_tree.get("parameters/playback")
 
+signal collided
```
Comment (demwilson) [inline Player.gd:24]: This is new to me. What is a signal do?

File: TileMap.gd [hunk 3]
```diff
@@ -36,6 +36,38 @@ func _ready():
 	overworld = get_parent()
 	randomize()
 	build_level()
+
+func _on_Player_collided(collision):
+	if collision.collider is TileMap:
+		var playerTilePos
+		if collision.normal.x == 1:
+			var remainderX = int(overworld.player.position.x) % 32
```
Comment (demwilson) [inline TileMap.gd]: What is the magic 32 number?

File: TileMap.gd [hunk 4]
```diff
@@ -36,6 +36,38 @@ func _ready():
...
+		else:
+			playerTilePos = self.world_to_map(overworld.player.position)
+
+		print("Player Position: " + str(playerTilePos))
+		print("Collision Normal: " + str(collision.normal))
+		playerTilePos -= collision.normal
+		print("Collision Object Position: " + str(playerTilePos))
+		var tile = collision.collider.get_cellv(playerTilePos)
+		print(str(tile))
+		if tile == 4:
```
Comment (demwilson) [inline TileMap.gd]: What is the magic 4 number?

File: TileMap.gd [hunk 5]
```diff
@@ -36,6 +36,38 @@ func _ready():
...
+			set_tile(playerTilePos.x, playerTilePos.y, Tile.Floor)
+		elif tile == 3:
```
Comment (demwilson) [inline TileMap.gd]: What is the magic 3 number?

**Review by demwilson — CHANGES_REQUESTED** [review]

File: Overworld.gd (hunk 1 above)
Comment (beserkzombie) [reply Overworld.gd]: takes a world position like (576, 256) and maps it to the tile based grid (18, 8)
formula being tile_pos = Vector2(pos.x / tile_size.x , pos.y /tile_size.y)
the built in method takes this into account and figures out the conversion

File: TileMap.gd (hunk 3 above)
Comment (beserkzombie) [reply TileMap.gd]: I will change this to tile_size

File: TileMap.gd (hunk 4 above)
Comment (beserkzombie) [reply TileMap.gd]: 4 is the tile enumeration value for Tile.Door, i think i should be able to just use Tile.Door

File: TileMap.gd (hunk 5 above)
Comment (beserkzombie) [reply TileMap.gd]: 3 is the tile enumeration value for Tile.Ladder, i think i should be able to just use Tile.Ladder

File: Player.gd (hunk 2 above)
Comment (beserkzombie) [reply Player.gd:24]: There is a tab on the right called "Node" and its where you can set up events/signals to be accessed by other objects.
declaring the signal adds it to the node list then you can go to the object listning for the signal and add the method for it to call by clicking on the node item.

File: Overworld.gd [hunk 6]
```diff
@@ -28,10 +28,13 @@ func place_player():
...
-	$CanvasLayer/TilePos.text = str(cpos)
-	var mpos = $TileMap.world_to_map(get_global_mouse_position())
-	$CanvasLayer/MousePos.text = str(mpos)
+#	#uncomment these lines to have player pos and mouse pos appear on canvas for testing
+#	$CanvasLayer/TilePos.visible = true
+#	$CanvasLayer/MousePos.visible = true
+#	var cpos = $TileMap.world_to_map($Player.position)
+#	$CanvasLayer/TilePos.text = str(cpos)
+#	var mpos = $TileMap.world_to_map(get_global_mouse_position())
+#	$CanvasLayer/MousePos.text = str(mpos)
```
Comment (demwilson) [inline Overworld.gd]: You should create a `debug` property in Global and have this execute if it's true.

Comment (beserkzombie) [reply Overworld.gd]: scope creepe

Comment (demwilson) [reply Overworld.gd]: Maybe a little, but I think it's worth while.

Comment (beserkzombie) [reply Overworld.gd]: worth

File: Player.gd (hunk 2 above)
Comment (demwilson) [reply Player.gd:24]: Cool.

**Review by demwilson — APPROVED** [review]

**Pull request merged**